Code/Tools/AssetProcessor/native/ui/BuilderInfoMetricsModel.cpp

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#include <ui/BuilderInfoMetricsModel.h>
+#include <ui/BuilderDataItem.h>
+#include <ui/BuilderData.h>
+#include <AssetBuilderSDK/AssetBuilderSDK.h>
+#include <QTime>
+
+namespace AssetProcessor
+{
+    QString DurationToQString(AZ::s64 durationInMs)
+    {
+        const AZ::s64 dayInMs = 86400000;
+        if (durationInMs < 0)
+        {
+            return "";
+        }
+
+        AZ::s64 dayCount = durationInMs / dayInMs;
+        QTime duration = QTime::fromMSecsSinceStartOfDay(durationInMs % dayInMs);
+        if (dayCount > 0)
+        {
+            return duration.toString("zzz' ms, 'ss' sec, 'mm' min, 'hh' hr, %1 day'").arg(dayCount);
+        }
+        else
+        {
+            if (!duration.isValid())
+            {
+                return "";
+            }
+            if (duration.hour() > 0)
+            {
+                return duration.toString("zzz' ms, 'ss' sec, 'mm' min, 'hh' hr'");
+            }
+            if (duration.minute() > 0)
+            {
+                return duration.toString("zzz' ms, 'ss' sec, 'mm' min'");
+            }
+            if (duration.second() > 0)
+            {
+                return duration.toString("zzz' ms, 'ss' sec'");
+            }
+            return duration.toString("zzz' ms'");
+        }
+    }
+
+    BuilderInfoMetricsModel::BuilderInfoMetricsModel(BuilderData* builderData, QObject* parent)
+        : QAbstractItemModel(parent)
+        , m_data(builderData)
+    {
+    }
+
+    void BuilderInfoMetricsModel::Reset()
+    {
+        beginResetModel();
+        m_data->Reset();
+        endResetModel();
+    }
+
+    void BuilderInfoMetricsModel::OnBuilderSelectionChanged(const AssetBuilderSDK::AssetBuilderDesc& builder)
+    {
+        // TODO: need to handle the "all builders" case
```

Comment:
which builder cases are not handled? 
or should this TODO be removed because all builders are handled?

Reply:
This should be kept. I paraphrased this confusing TODO comment.

In my next move, the BuilderListModel will turn from a simple list to a tree where all builders are children of a root item, which is called "All Builders" and, as its display name suggests, gives metrics that are accumulated from all the builders. On clicking this item, the BuilderInfoMetricsModel should be able to display the metrics aggregated from all builders' metrics. The data is collected in BuilderData, but this method, OnBuilderSelectionChanged, currently has no way to pass in this "All Builders" option and act accordingly. This is thus still a valid TODO comment that is expected to disappear in my next PR.